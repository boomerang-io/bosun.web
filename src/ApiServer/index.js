import { Server, Serializer, Model } from "miragejs";
import { inflections } from "inflected";
import uuid from "uuid/v4";
import { serviceUrl } from "Config/servicesConfig";
import * as fixtures from "./fixtures";

export function startApiServer({ environment = "test", timing = 0 } = {}) {
  inflections("en", function (inflect) {
    // Prevent pluralization bc our apis are weird
    inflect.irregular("policy", "policy");
    inflect.irregular("violation", "violation");
    inflect.irregular("template", "template");
    inflect.irregular("insights", "insights");
  });

  return new Server({
    environment,
    // Load in mock data
    fixtures,
    // Return the data as is, don't add a root key
    serializers: {
      application: Serializer.extend({
        root: false,
        embed: true,
      }),
    },
    // Register the data as a model so we can use the schema
    models: {
      templates: Model,
      policies: Model,
    },

    routes() {
      // Control how long the responses take to resolve
      this.timing = timing;

      // Allow unhandled requests on the current domain to pass through
      this.passthrough();

      this.get("/info", () => []);

      /**
       * Simple GET of static data
       */
      this.get(serviceUrl.getUserProfile(), (schema) => {
        return schema.db.profile[0];
      });
      this.get(serviceUrl.getCicdNavigation({}), (schema, request) => {
        return schema.db.cicdNavigation.map((item) => {
          if (typeof item.link === "string") {
            item.link = item.link.replace(":teamId", request.queryParams.teamId);
          }

          return item;
        });
      });

      this.get(serviceUrl.getPlatformNavigation(), (schema) => {
        return schema.db.navigation[0];
      });

      this.get(serviceUrl.getTeams(), (schema) => {
        return schema.db.teams;
      });

      this.get(serviceUrl.getPolicy({ policyId: ":policyId" }), (schema, request) => {
        let { policyId } = request.params;
        let activePolicy = schema.policies.find(policyId);
        return activePolicy;
      });

      this.get(serviceUrl.getValidateInfo({ policyId: ":policyId" }), (schema) => {
        return schema.db.validateInfo;
      });

      this.get(serviceUrl.getTemplates(), (schema) => {
        return schema.db.templates;
      });

      this.get(serviceUrl.getPolicyOverview(), (schema, request) => {
        let { teamId } = request.queryParams;
        return schema.db.policies.filter((policy) => policy.teamId === teamId);
      });

      this.get(serviceUrl.getViolationsOverview(), (schema) => {
        return schema.db.violations;
      });

      this.get(serviceUrl.getInsightsOverview(), (schema) => {
        return schema.db.insights;
      });

      this.post(serviceUrl.postCreatePolicy(), (schema, request) => {
        let body = JSON.parse(request.requestBody);
        return schema.policies.create({ ...body, id: uuid() });
      });

      this.post(serviceUrl.postCreatePolicyTemplate(), (schema, request) => {
        let body = JSON.parse(request.requestBody);
        return schema.templates.create({ ...body, id: uuid() });
      });

      this.patch(serviceUrl.patchUpdatePolicy({ policyId: ":policyId" }), (schema, request) => {
        let { policyId } = request.params;
        let body = JSON.parse(request.requestBody);
        let activePolicy = schema.policies.find(policyId);
        activePolicy.update(body);
        return schema.policies.all();
      });

      this.patch(serviceUrl.patchUpdatePolicyTemplate({ templateId: ":templateId" }), (schema, request) => {
        let { templateId } = request.params;
        let body = JSON.parse(request.requestBody);
        let activeTemplate = schema.templates.find(templateId);
        activeTemplate.update(body);
        return schema.templates.all();
      });

      this.delete(serviceUrl.deletePolicy({ policyId: ":policyId" }), (schema, request) => {
        let { policyId } = request.params;
        let activePolicy = schema.policies.find(policyId);
        activePolicy.destroy({ id: policyId });
        return schema.policies.all();
      });
    },
  });
}
