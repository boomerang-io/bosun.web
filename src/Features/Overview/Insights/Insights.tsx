import React from "react";
import Graph from "./Graph";
import Info from "./Info";
import { getLineChartData } from "./helpers/formatLineChartData";
import { PolicyData, InsightsData, Violation, Info as InfoProps, ChartsData } from "Types";
import styles from "./insights.module.scss";

type Props = {
  insights: InsightsData[];
  policies: PolicyData[];
  violations: Violation[];
};

function Insights({ insights, violations, policies }: Props) {
  const formatedData: ChartsData = getLineChartData(insights);
  let gatesCount = 0;

  const gatesData: InfoProps = {
    type: "gates",
    title: "Current Gates",
    content: "Software never is perfect and it wonâ€™t be. But is that a license to create garbage?",
    count: policies.reduce((acc, policy) => (gatesCount += policy.stages?.length ?? 0), 0),
  };
  const violationsData: InfoProps = {
    type: "violations",
    title: "Current Violations",
    content: "Code will set you free. Think of this as an opportunity for better code. Drive towards zero.",
    count: violations.reduce((acc, violation) => (acc += violation?.nbrViolations ?? 0), 0),
  };

  return (
    <section className={styles.container}>
      <h2 className={styles.title}>Insights</h2>
      <div className={styles.contentContainer}>
        <div className={styles.tilesContainer}>
          <Info info={gatesData} />
          <Info info={violationsData} />
        </div>
        <div className={styles.graphsContainer}>
          <Graph formatedData={formatedData} />
        </div>
      </div>
    </section>
  );
}

export default Insights;
