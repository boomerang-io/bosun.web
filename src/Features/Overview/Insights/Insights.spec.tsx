import React from "react";
import Insights from "./index";
import { insights, policies as policiesData, violations as violationsData } from "ApiServer/fixtures";

const violations = [violationsData[0]];
const policies = [policiesData[1]];

describe("Insights --- Snapshot", () => {
  it("Capturing Snapshot of Insights", async () => {
    const { baseElement } = (global as any).rtlRender(
      <Insights insights={insights} policies={policies} violations={violations} />
    );
    expect(baseElement).toMatchSnapshot();
  });
});
