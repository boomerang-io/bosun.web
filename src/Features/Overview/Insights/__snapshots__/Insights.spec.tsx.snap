// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Insights --- Snapshot Capturing Snapshot of Insights 1`] = `
<body
  id="app"
>
  <div>
    <section
      class="container"
    >
      <h2
        class="title"
      >
        Insights
      </h2>
      <div
        class="contentContainer"
      >
        <div
          class="tilesContainer"
        >
          <div
            class="bx--tile"
            data-testid="tile-info"
            style="display: flex; width: 100%; height: 10rem; align-items: center; padding: 0px; margin-bottom: 1rem; margin-left: 0px;"
          >
            <div
              class="counter"
              style="color: rgb(4, 124, 192);"
            >
              1
            </div>
            <div
              class="divider"
            />
            <div
              class="text"
            >
              <h3
                class="title"
              >
                Current Gates
              </h3>
              <p
                class="content"
              >
                Software never is perfect and it wonâ€™t be. But is that a license to create garbage?
              </p>
            </div>
          </div>
          <div
            class="bx--tile"
            data-testid="tile-info"
            style="display: flex; width: 100%; height: 10rem; align-items: center; padding: 0px; margin-bottom: 1rem; margin-left: 1rem;"
          >
            <div
              class="counter"
              style="color: rgb(255, 170, 157);"
            >
              1
            </div>
            <div
              class="divider"
            />
            <div
              class="text"
            >
              <h3
                class="title"
              >
                Current Violations
              </h3>
              <p
                class="content"
              >
                Code will set you free. Think of this as an opportunity for better code. Drive towards zero.
              </p>
            </div>
          </div>
        </div>
        <div
          class="graphsContainer"
        >
          <div
            class="bx--tile"
            style="display: flex; flex-direction: column; align-items: spaceBetween; min-height: 10rem;"
          >
            <div
              class="header"
            >
              <h3
                class="title"
              >
                Violations Trend
              </h3>
              <div
                class="filters"
              />
            </div>
            <div
              class="container"
              data-testid="insights-graph"
            >
              <div
                style="display: flex; align-items: center; width: 100%;"
              >
                <div
                  class="recharts-responsive-container"
                  style="width: 99%; height: 100%;"
                >
                  <div
                    style="position: absolute; width: 0px; height: 0px; visibility: hidden; display: none;"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </section>
  </div>
</body>
`;
